### ISSUE TITLE
Divergent tensor values in YOLOv8 forward pass between model.train(True) and model.train(False), with subsequent decline in performance in inference mode

### ISSUE THREAD

**Issue by PabloMessina** [ultralytics/ultralytics#7118]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Other

### Bug

I noticed an issue while training a model whose core visual processing module is YOLOv8, from which I extract visual features feeding other parts of the model. The issue was that the metrics I was getting in validation where consistently lower than the metrics I was getting in training, **even when the same data is used in both**. That is, using exactly the same dataloader, training the model on the very same data that it is evaluated on, the metrics during training were higher than the metrics during validation. This was very strange, and it prompted me to spend a lot of time on debugging, trying different things, from freezing parameters to deactivating automatic mixed precision, deactivating backpropagation, sanity checking my metric computation code, making sure exactly the same dataloader is used in both cases, etc. Finally, I was able to narrow down the problem to its actual cause: in inference mode I execute model.train(False), whereas in training mode I execute  model.train(True). So I decided to run model.train(True) in validation as well, and voila, the performance in validation was high and exactly the same as the performance I was getting in training mode. This means that there must be something going behind the scenes inside YOLOv8 that changes the computations in inference mode, thus degrading performance. So I decided to log the values of the tensors that YOLOv8 computes in sequence during a forward pass in both settings (training and evaluation), and it turns out that very early on the values start to diverge. Below some screenshots:

### With model.train(True):

<img width="509" alt="image" src="https://github.com/ultralytics/ultralytics/assets/6187545/bbbc4e30-147c-4adf-8043-6cb8747715b0">

### With model.train(False)
<img width="578" alt="image" src="https://github.com/ultralytics/ultralytics/assets/6187545/7d1a5c15-1f8a-4b76-b120-34e84242c022">

When I run the model with model.train(True), the outputs make a lot of sense, and the model behaves as expected. But if I run it with model.train(False) (and with everything else being equal, the same input image, etc.), the performance degrades, because the tensor values change and therefore subsequent layers that rely on those values get confused.

What's going on, and how can it be fixed?

### Environment

Ultralytics YOLOv8.0.58 🚀 Python-3.10.10 torch-1.13.1+cu117 CUDA:0 (GeForce RTX 3090, 24267MiB)
Setup complete ✅ (20 CPUs, 125.6 GB RAM, 579.2/681.1 GB disk)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@PabloMessina hello! Thanks for the detailed description of your issue. The behavior you're observing where performance differs between `model.train(True)` and `model.train(False)` is typically related to the presence of layers or modules within YOLOv8 that have different behaviors during training and inference, such as batch normalization and dropout layers.

Batch normalization layers, in particular, maintain running estimates of mean and variance during training. During inference (when `train(False)` is set), those estimates are used to normalize the batch, rather than the statistics of the current batch. This can lead to the type of divergence in metrics you're seeing if, for example:

1. The running estimates do not represent the data well (e.g., due to insufficient training or highly varied data).
2. The model has been recently initialized and the running estimates haven't stabilized.
3. You're not using the `.eval()` method which correctly sets the model to inference mode beyond just the training flag.

A good practice is to make sure the model is in evaluation mode by calling `model.eval()` during validation/testing, which correctly prepares the model for inference by properly setting batch normalization and dropout layers, among other things.

If you verify that all this is correctly set and still observe degrading performance, there may be other factors at play. For instance, ensuring that the data preprocessing and augmentation steps are consistent between training and validation can also be crucial, as discrepancies there can impact performance.

You might also consider reviewing the details of the custom components you have and how they interact with YOLOv8. If these custom parts rely heavily on batch statistics, the switch to inference mode could disproportionately affect them.

If the issue persists after double-checking these areas, please provide more details, and we'll look into this further. Thanks for your support and for contributing to the YOLO community! 🚀

**Comment by PabloMessina** [comment]
@glenn-jocher Thanks for the prompt reply.

> The behavior you're observing where performance differs between model.train(True) and model.train(False) is typically related to the presence of layers or modules within YOLOv8 that have different behaviors during training and inference, such as **batch normalization and dropout layers**

Could you please indicate which layers are you referring to specifically? I don't see any Dropout nor Batch Normalization layers in YOLOv8's architecture summary (see screenshot below):

<img width="488" alt="image" src="https://github.com/ultralytics/ultralytics/assets/6187545/ba46fa30-4c06-450d-861d-c60c04b4832d">

> A good practice is to make sure the model is in evaluation mode by calling model.eval() during validation/testing, which correctly prepares the model for inference by properly setting batch normalization and dropout layers, among other things.

I understand that model.train(False) already invokes model.eval() behind the scenes, but anyways, I explicitly called model.eval() and it made no difference, I got the same result.

But anyways, I'm just testing YOLOv8 with a single example, i.e., a batch of size 1. That's why in the tensor shapes in the screenshots I shared in my previous comment you see (1, 3, 416, 416), (1, 64, 208, 208), etc. Using this single example, when I run the model with model.train(True), the outputs are great. But when I run it with model.train(False) (or model.eval(), which is the same thing), the performance degrades, because the tensor values change. Is there anyway to force YOLOv8 to compute the same values?

Can you please point me to the specific layer that is causing the changes in computation?

**Comment by glenn-jocher** [comment]
@PabloMessina, YOLOv8 uses certain normalization techniques within its architecture that behave differently during training and inference. These are not traditional dropout layers but include normalization techniques possibly within custom modules or within certain layers like convolutional ones.

With single-example batches, you're effectively testing the model in a different regime from both training and normal inference because normalization statistics (batch statistics versus running estimates) differ more when batch sizes are small.

The fact that `model.eval()` did not resolve the discrepancy suggests that another feature linked to training mode is at play. Without detailed context, it isn't straightforward to pinpoint the exact layer or mechanism responsible just by looking at the model summary.

The `DetectionModel` you mentioned uses the `Model` class from the YOLOv8 framework, which handles the transition between training and inference modes. The behavior during these modes comes from the PyTorch API and our custom implementations. You can inspect the forward methods within the model architecture where these transitions could change behavior.

If you wish to prevent any change in behavior between training and inference for debugging purposes, you might temporarily alter the forward passes of normalization layers to use training statistics even during inference. However, this is not recommended for general use and should only be done with caution and a clear understanding of the implications.

Since it is not advisable to share specific source code lines in this medium, you should refer to the model definition files in YOLOv8's repository and trace the forward passes from there. That should help identify any special conditional logic tied to the model's training state.
